compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,242 @@
+          //   refinedInfo = TypeRef(..., Int)
+          // )
+          type RefinedInfo = (core.Names.Name, Type)
+          def flatten(tpe: Type, acc: List[RefinedInfo]): (Type, List[RefinedInfo]) = tpe match {
+            case RefinedType(parent, name, info) =>
+              flatten(parent, acc :+ (name, info))
+            case _ =>
+              (tpe, acc)
+          }
+
+          // flatten parent types to list
+          // e.g. `X with Y with Z { refined }`
+          // RefinedType(parent = AndType(X, AndType(Y, Z)), ...)
+          // => List(X, Y, Z)
+          def flattenParent(parent: Type): List[s.Type] = parent match {
+            case AndType(tp1, tp2) =>
+              flattenParent(tp1) ++ flattenParent(tp2)
+            case _ => List(loop(parent))
+          }
+
+          val (parent, refinedInfos) = flatten(rt, List.empty)
+          val stpe = s.WithType(flattenParent(parent))
+
+          // Create dummy symbols for refinements
+          // since there's no way to retrieve symbols of refinements from RefinedType at this moment.
+          val decls = for (name, info) <- refinedInfos
+            yield newSymbol(sym, name, Flags.EmptyFlags, info)
+          val sdecls = decls.sscope(using LinkMode.HardlinkChildren)
+          s.StructuralType(stpe, Some(sdecls))
+
+        case rec: RecType =>
+          loop(rec.parent) // should be handled as RefinedType
+
+        // repeated params: e.g. `Int*`, which is the syntax sugar of
+        // `Seq[Int] @Repeated` (or `Array[Int] @Repeated`)
+        // See: Desugar.scala and TypeApplications.scala
+        case AnnotatedType(AppliedType(_, targs), annot)
+          if (annot matches defn.RepeatedAnnot) && (targs.length == 1) =>
+          val stpe = loop(targs(0))
+          s.RepeatedType(stpe)
+
+        case ann: AnnotatedType if ann.annot.symbol.info.isInstanceOf[ClassInfo] =>
+          def flatten(tpe: Type, annots: List[Annotation]): (Type, List[Annotation]) = tpe match
+            case AnnotatedType(parent, annot) if annot.symbol.info.isInstanceOf[ClassInfo] =>
+              flatten(parent, annot +: annots)
+            case other => (other, annots)
+
+          val (parent, annots) = flatten(ann, List.empty)
+          val sparent = loop(parent)
+          val sannots = annots.map(a =>
+            s.Annotation(loop(a.symbol.info.asInstanceOf[ClassInfo].selfType))
+          )
+          s.AnnotatedType(sannots, sparent)
+
+        case app @ AppliedType(tycon, args) =>
+          val sargs = args.map(loop)
+          loop(tycon) match
+            case ref: s.TypeRef => ref.copy(typeArguments = sargs)
+            // is there any other cases?
+            case _ => s.Type.Empty
+
+        case and: AndType =>
+          def flatten(child: Type): List[Type] = child match
+            case AndType(ct1, ct2) => flatten(ct1) ++ flatten(ct2)
+            case other => List(other)
+          val stpes = flatten(and).map(loop)
+          s.WithType(stpes)
```

Comment:
> My comment about IntersectionType was in case some tooling makes an assumption that it implies the definition must be a Java type, but maybe this is not the case in practise. It is true that a Java intersection type has different erasure to both a scala 3 intersection type or a Scala 2 WithType, so it might be worth treating separately

@bishabosha  I don't really think we need to treat them separately. We already have separate sections for Java and Scala in https://scalameta.org/docs/semanticdb/specification.html#type-2 and the reason for it is that same type in Scala file might mean something different. There is no need to model intersection types as WithType, since this is leaking the implementation details. When consuming semanticdb we only care about it being an intersection and we already know which file we are in. There is never an assumption that any type is java or scala specific. Semanticdb is intended to represent a lot of different languages (though it's Scala and Java only currently)

Reply:
agreed